src/clarity_vm/special.rs

```diff
@@ -1011,6 +1264,18 @@ pub fn handle_contract_call_special_cases(
             args,
             result,
         );
+    } else if *contract_id == boot_code_id(POX_3_NAME, global_context.mainnet)
+        && global_context.database.get_pox_3_activation_height()
+            >= global_context.database.get_current_burnchain_block_height()
+    {
+        return handle_pox_v3_api_contract_call(
+            global_context,
+            sender,
+            contract_id,
+            function_name,
+            args,
+            result,
+        );
```

Comment:
I think it's impossible for the second part of the if statement to fail (i.e., you can't get a contract-call to pox-3 before the activation height), but just in case, I think something like the following would be better:

```suggestion
    } else if *contract_id == boot_code_id(POX_3_NAME, global_context.mainnet) {
        if global_context.database.get_pox_3_activation_height()
            < global_context.database.get_current_burnchain_block_height() {
            warn!("PoX-3 contract invoked before PoX-3 activation height");
            return Err(Error::Runtime(RuntimeErrorType::DefunctPoxContract, None));
        }
        return handle_pox_v3_api_contract_call(
            global_context,
            sender,
            contract_id,
            function_name,
            args,
            result,
        );
```

This way, any method that is invoked in PoX-3 before the activation height will runtime error.

Reply:
Yesterday I was able to boot and transition into epoch2.4 (local regtest setup). After a recent commit from last night, I'm not longer able to. It panics at the epoch2.4 transition:
```rs
if global_context.database.get_current_burnchain_block_height() // <-- evaluates to 109
    < global_context.database.get_pox_3_activation_height() //  <-- evaluates to 110
 {
    warn!("PoX-3 contract invoked before PoX-3 activation height");
    return Err(Error::Runtime(RuntimeErrorType::DefunctPoxContract, None));
}
```

Unclear to me if this is a bug in my epoch2.4 config, or a bug in this condition. Any idea?

Reply:
Error log + stack trace:
```
INFO [1683208876.861945] [src/chainstate/stacks/db/blocks.rs:4871] [miner-block-http://0.0.0.0:20443] Applying epoch transition, new_epoch_id: 2.4, old_epoch_id: 2.3
WARN [1683209476.541722] [src/clarity_vm/special.rs:1272] [miner-block-http://0.0.0.0:20443] PoX-3 contract invoked before PoX-3 activation height
ERRO [1683209476.542718] [testnet/stacks-node/src/main.rs:58] [miner-block-http://0.0.0.0:20443] Process abort due to thread panic: panicked at 'Failed to set burnchain parameters in PoX-3 contract: Interpreter(Runtime(DefunctPoxContract, None))', src/clarity_vm/clarity.rs:1260:22
ERRO [1683209476.584315] [testnet/stacks-node/src/main.rs:60] [miner-block-http://0.0.0.0:20443] Panic backtrace:    0: backtrace::backtrace::libunwind::trace
             at /Users/matt/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.67/src/backtrace/libunwind.rs:93:5
      backtrace::backtrace::trace_unsynchronized
             at /Users/matt/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.67/src/backtrace/mod.rs:66:5
      backtrace::backtrace::trace
             at /Users/matt/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.67/src/backtrace/mod.rs:53:14
      backtrace::capture::Backtrace::create
             at /Users/matt/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.67/src/capture.rs:176:9
      backtrace::capture::Backtrace::new
             at /Users/matt/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.67/src/capture.rs:140:22
   1: stacks_node::main::{{closure}}
             at testnet/stacks-node/src/main.rs:59:18
   2: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/alloc/src/boxed.rs:2002:9
      std::panicking::rust_panic_with_hook
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/panicking.rs:692:13
   3: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/panicking.rs:579:13
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/sys_common/backtrace.rs:137:18
   5: rust_begin_unwind
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/panicking.rs:575:5
   6: core::panicking::panic_fmt
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/panicking.rs:64:14
   7: core::result::unwrap_failed
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/result.rs:1790:5
   8: core::result::Result<T,E>::expect
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/result.rs:1069:23
      blockstack_lib::clarity_vm::clarity::ClarityBlockConnection::initialize_epoch_2_4::{{closure}}::{{closure}}
             at src/clarity_vm/clarity.rs:1251:56
      blockstack_lib::clarity_vm::clarity::ClarityBlockConnection::as_transaction
             at src/clarity_vm/clarity.rs:1332:17
   9: blockstack_lib::clarity_vm::clarity::ClarityBlockConnection::initialize_epoch_2_4::{{closure}}
             at src/clarity_vm/clarity.rs:1230:48
      blockstack_lib::clarity_vm::clarity::ClarityBlockConnection::initialize_epoch_2_4
             at src/clarity_vm/clarity.rs:1145:9
  10: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::process_epoch_transition
             at src/chainstate/stacks/db/blocks.rs:4981:46
  11: blockstack_lib::chainstate::stacks::db::blocks::<impl blockstack_lib::chainstate::stacks::db::StacksChainState>::setup_block
             at src/chainstate/stacks/db/blocks.rs:5818:13
  12: blockstack_lib::chainstate::stacks::miner::<impl blockstack_lib::chainstate::stacks::StacksBlockBuilder>::epoch_begin
             at src/chainstate/stacks/miner.rs:2120:13
  13: blockstack_lib::chainstate::stacks::miner::<impl blockstack_lib::chainstate::stacks::StacksBlockBuilder>::build_anchored_block
             at src/chainstate/stacks/miner.rs:2396:13
  14: stacks_node::neon_node::BlockMinerThread::run_tenure
             at testnet/stacks-node/src/neon_node.rs:1909:44
  15: stacks_node::neon_node::RelayerThread::block_miner_thread_try_start::{{closure}}
             at testnet/stacks-node/src/neon_node.rs:3075:28
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/sys_common/backtrace.rs:121:18
  16: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/thread/mod.rs:558:17
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/panic/unwind_safe.rs:271:9
      std::panicking::try::do_call
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/panicking.rs:483:40
      std::panicking::try
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/panicking.rs:447:19
      std::panic::catch_unwind
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/panic.rs:140:14
      std::thread::Builder::spawn_unchecked_::{{closure}}
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/thread/mod.rs:557:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/ops/function.rs:250:5
  17: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/alloc/src/boxed.rs:1988:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/alloc/src/boxed.rs:1988:9
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/sys/unix/thread.rs:108:17
  18: __pthread_deallocate

Dumping core for pid 50958
```

Reply:
Rewinding the call stack I get to this line where `burn_tip_height` is `110`:
```rs
        let (applied_epoch_transition, mut tx_receipts) =
            StacksChainState::process_epoch_transition(&mut clarity_tx, burn_tip_height)?;
```

Seems like a mismatch between the burn height here and the one returned later on by `global_context.database.get_current_burnchain_block_height()`

Reply:
Yeah -- `get_current_burnchain_block_height()` will return the _parent's_ block height.

Reply:
(note, this is intended behavior, because that's the block height at which you can query for burnchain information)

Reply:
But I think this does mean this code here needs to be updated -- because the `pox-3` contract does need to be invoked during initialization _before_ the activation height is reached.